File: sdk/php/src/SecretKey.php
<?php

declare(strict_types=1);

namespace HideProtocol;

/**
 * A secret key. The bytes stay inside the native library and are never exposed
 * to PHP — there is deliberately no accessor for them.
 *
 * Release it with close(); it is also released when the object is collected.
 */
final class SecretKey
{
    private ?\FFI\CData $handle;

    private function __construct(\FFI\CData $handle)
    {
        $this->handle = $handle;
    }

    public static function generate(): self
    {
        $ffi = Binding::ffi();
        $handle = $ffi->new('void *');
        $public = Binding::emptyBuffer();

        try {
            Binding::check($ffi->hide_keypair_generate(\FFI::addr($handle), \FFI::addr($public)));
        } finally {
            $ffi->hide_buffer_free(\FFI::addr($public));
        }

        return new self($handle);
    }

    /** Opens a key file. A protected key without its passphrase fails. */
    public static function open(string $bytes, ?string $passphrase = null): self
    {
        $ffi = Binding::ffi();
        $handle = $ffi->new('void *');
        $data = Binding::bytes($bytes);
        $secret = Binding::cString($passphrase, 'passphrase');

        try {
            Binding::check($ffi->hide_secret_key_open(
                $ffi->cast('uint8_t *', $data),
                strlen($bytes),
                $secret,
                \FFI::addr($handle),
            ));
        } finally {
            Binding::release($data);
            Binding::release($secret);
        }

        return new self($handle);
    }

    public function publicKey(): string
    {
        $handle = $this->alive();
        $out = Binding::emptyBuffer();
        Binding::check(Binding::ffi()->hide_secret_key_public($handle, \FFI::addr($out)));

        return Binding::take($out);
    }

    /**
     * Seals this key with a passphrase, for writing to disk.
     *
     * A forgotten passphrase cannot be recovered: there is no escrow.
     */
    public function protect(string $passphrase): string
    {
        $handle = $this->alive();
        if (strlen($passphrase) < Hide::MIN_PASSPHRASE_LEN) {
            throw new \InvalidArgumentException(
                'the passphrase must be at least ' . Hide::MIN_PASSPHRASE_LEN . ' characters'
            );
        }

        $out = Binding::emptyBuffer();
        $secret = Binding::cString($passphrase, 'passphrase');
        try {
            Binding::check(Binding::ffi()->hide_secret_key_protect($handle, $secret, \FFI::addr($out)));
        } finally {
            Binding::release($secret);
        }

        return Binding::take($out);
    }

    public function close(): void
    {
        if ($this->handle !== null) {
            Binding::ffi()->hide_secret_key_free($this->handle);
            $this->handle = null;
        }
    }

    public function isClosed(): bool
    {
        return $this->handle === null;
    }

    /** @internal */
    public function alive(): \FFI\CData
    {
        if ($this->handle === null) {
            throw new \InvalidArgumentException('this key has been closed');
        }

        return $this->handle;
    }

    public function __destruct()
    {
        $this->close();
    }

    /**
     * var_dump() and print_r() go through this, so it must never render key
     * material — not even a fingerprint derived from it.
     *
     * @return array{state: string}
     */
    public function __debugInfo(): array
    {
        return ['state' => $this->handle === null ? 'closed' : 'open'];
    }

    public function __toString(): string
    {
        return 'HideProtocol\SecretKey(' . ($this->handle === null ? 'closed' : 'open') . ')';
    }
}

File: sdk/dotnet/HideProtocol/SecretKey.cs
namespace HideProtocol;

/// <summary>
/// A secret key. The bytes stay inside the native library and are never exposed
/// to managed code. Dispose it when finished.
/// </summary>
public sealed unsafe class SecretKey : IDisposable
{
    private IntPtr _handle;

    private SecretKey(IntPtr handle) => _handle = handle;

    public static SecretKey Generate()
    {
        IntPtr handle = IntPtr.Zero;
        HideBuffer publicKey = Native.hide_buffer_empty();
        Interop.Check(Native.hide_keypair_generate(&handle, &publicKey));
        Interop.Take(ref publicKey);
        return new SecretKey(handle);
    }

    /// <summary>Loads a key file. A protected key without its passphrase fails.</summary>
    public static SecretKey Open(ReadOnlySpan<byte> bytes, string? passphrase = null)
    {
        IntPtr handle = IntPtr.Zero;
        using Utf8String pass = Utf8String.Create(passphrase, nameof(passphrase));
        fixed (byte* data = bytes)
        {
            Interop.Check(Native.hide_secret_key_open(
                data, (nuint)bytes.Length, pass.Pointer, &handle));
        }

        return new SecretKey(handle);
    }

    public byte[] PublicKey()
    {
        IntPtr handle = Alive();
        HideBuffer output = Native.hide_buffer_empty();
        Interop.Check(Native.hide_secret_key_public(handle, &output));
        return Interop.Take(ref output);
    }

    /// <summary>
    /// Seals this key with a passphrase, for writing to disk. A forgotten
    /// passphrase cannot be recovered: there is no escrow.
    /// </summary>
    public byte[] Protect(string passphrase)
    {
        ArgumentNullException.ThrowIfNull(passphrase);
        IntPtr handle = Alive();
        if (passphrase.Length < Hide.MinPassphraseLength)
        {
            throw new ArgumentException(
                $"the passphrase must be at least {Hide.MinPassphraseLength} characters",
                nameof(passphrase));
        }

        using Utf8String pass = Utf8String.Create(passphrase, nameof(passphrase));
        HideBuffer output = Native.hide_buffer_empty();
        Interop.Check(Native.hide_secret_key_protect(handle, pass.Pointer, &output));
        return Interop.Take(ref output);
    }

    internal IntPtr Alive() =>
        _handle == IntPtr.Zero ? throw new ObjectDisposedException(nameof(SecretKey)) : _handle;

    public void Dispose()
    {
        if (_handle != IntPtr.Zero)
        {
            Native.hide_secret_key_free(_handle);
            _handle = IntPtr.Zero;
        }

        GC.SuppressFinalize(this);
    }

    ~SecretKey() => Dispose();

    // Never render key material, not even a fingerprint of it.
    public override string ToString() =>
        $"HideProtocol.SecretKey ({(_handle == IntPtr.Zero ? "disposed" : "open")})";
}
